### PR TITLE
Fix #3938: adapt prefix to ease prefix inference

### LINKED ISSUE #3938: Erroneous unreachable warning with path dependant type
```scala
class Foo {
  val bar = new Bar
  class Bar {
    sealed abstract class A
    case class B() extends A
    case class C() extends A
  }
}

class Test {
  val foo = new Foo
  import foo.bar._

  def test(a: A) = {
    a match {
      case B() => 1
      case _ => 2 // unreachable code
    }
  }
}
```

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #3938: adapt prefix to ease prefix inference

### CODE REVIEW

**Comment by liufengyun** [comment]
@odersky : we are doing _prefix inference_ in exhaustivity check, this PR does some magic in order for inference to work smoothly. Could you have a look and see if it makes sense?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -638,15 +640,31 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
       }
     }
 
+    // Fix subtype checking for child instantiation,
+    // such that `Foo(Test.this.foo) <:< Foo(Foo.this)`
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I could not correlate the claim `Foo(Test.this.foo) <:< Foo(Foo.this)` in i3939.scala with the code. In i3938, foo is not a module, so how does the `removeThisType` affect it?

Also, note that if we do have a 

     object Foo

then we have already `Foo.type <:< Foo.this` and `Foo.this <:< Foo.type` by the rules for `ThisType` in TypeComparer's firstTry and secondTry methods. So I am sure why the code would be necessesary.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Thanks for the clarification. The correlation is that in the `else` branch, it throws away of the `ThisType`. For the `module` case, I'll try if I can throw it away.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I just checked, the patch for `module` cannot be removed, as the subtyping for `Foo.type <:< Foo.this` only works if it's static owner:

```Scala
          case tp1: NamedType if cls2.is(Module) && cls2.eq(tp1.widen.typeSymbol) =>
            cls2.isStaticOwner ||
            isSubType(tp1.prefix, cls2.owner.thisType) ||
            secondTry(tp1, tp2)
```

The test case `tests/patmat/3455.scala` shows that we need the patch:

```Scala
trait AxisCompanion {
   sealed trait Format
   object Format {
      case object Decimal extends Format
      case object Integer extends Format
   }
}
object Axis extends AxisCompanion
class Axis {
   import Axis._
   def test( f: Format ) = f match {
      case Format.Integer => "Int"
   }
}
```

**Pull request merged**